### ISSUE TITLE
How do I set the value of the "coll_tuned_alltoall_algorithm_segmentsize" parameter?

### ISSUE THREAD

**Issue by gukejun1** [open-mpi/ompi#11386]
For ompi-v4.0.5, the "-mac coll_tuned_alltoall_algorithm 0 " parameter can be added when the mpirun is running. The value range of the "-mac coll_tuned_alltoall_algorithm 0 " parameter is [0, 1, 2, 3, 4, 5]. However, the values of the "coll_tuned_alltoall_algorithm_segmentsize, coll_tuned_alltoall_algorithm_tree_fanout, and coll_tuned_alltoall_algorithm_chain_fanout " parameters are as follows: How do you set them up? The MPI runtime is optimal because the values of these three are adjusted. Can you help me answer this question?

**Comment by gukejun1** [comment]
`mpirun  -np 16   -hostfile hostfile -mca  coll_tuned_alltoall_algorithm 9 -mca coll_tuned_alltoall_algorithm_segmentsize 0 -mca coll_tuned_alltoall_algorithm_tree_fanout 4  -mca coll_tuned_alltoall_algorithm_chain_fanout 4 collective/mpi_osu_alltoall`  Like this, how do these three values set?

**Comment by gkatev** [comment]
Hi @gukejun1, AFAIK you also have to add `--mca coll_tuned_use_dynamic_rules true` for any of these parameters to take effect. Can you try with this addition?

**Comment by bosilca** [comment]
A trick I'm using for configuring the collectives per application is that I put them into an MCA file in the same directory as the application, and then I add the file to `mca_param_files` via the `--mca`.

**Comment by gukejun1** [comment]
> Hi @gukejun1, AFAIK you also have to add `--mca coll_tuned_use_dynamic_rules true` for any of these MCA parameters to take effect. You also might want something different than 0 for the segmentsize, as 0 means no segmentation (e.g. `--mca coll_tuned_alltoall_algorithm_segmentsize 64K`). Though keep in mind that some algorithms might not support segmentation. But that wasn't your question? Does the command line that you posted work?

Yes, I know this "segmentsize 0" means no segmentsize. The command line posted work. I'd like to know what'coll_tuned_alltoall_algorithm_segmentsize' range is? Is the value range of the int type ranging from - 2147483648 to 2147483647? Are the values of coll_tuned_alltoall_algorithm_tree_fanout and coll_tuned_alltoall_algorithm_chain_fanout the same? Do you have any suggestions for their values? Because I want to tune them to get the best latency.

**Comment by gkatev** [comment]
Yes it seems like that's the range for the segment size. The MCA parameter is registered here, and it's an int: https://github.com/open-mpi/ompi/blob/88d8972a4085f6492dc38786c925f9dec446d49e/ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c#L89-L95

Whether the algorithm will actually make use of every/any (positive) value in the range will depend on its actual implementation -- I suspect that as long as the segment size is less than the message size and the algorithm supports segmenting, segmenting will take place.

I don't personally have much experience with alltoall; I would suggest determining the best one for you experimentally. I can point you to some code if it's helpful. Assuming you've set an algorithm, the choice will be made here: https://github.com/open-mpi/ompi/blob/88d8972a4085f6492dc38786c925f9dec446d49e/ompi/mca/coll/tuned/coll_tuned_alltoall_decision.c#L155-L184

These algorithms are implemented in [coll/base/coll_base_alltoall.c](https://github.com/open-mpi/ompi/blob/eb87378cd8c55c08fde666c03a04aa2811da395a/ompi/mca/coll/base/coll_base_alltoall.c)

But also seeing the decision switch above, it appears that the fanout and segsize parameters are not passed to any algorithm. So none of the supported Alltoall algorithms support them (?) (or at least they are not configurable). Someone with more experience should confirm this.

**Comment by gukejun1** [comment]
> OK. Are there any requirements on the size of the "col_tuned_alltoall_algorithm_tree_fanout" and "col_tuned_alltoall_algorithm_chain_fanout" parameters? Is it related to the size of the message? What's the range setting?

**Comment by gukejun1** [comment]
Okay, got it. Thank you very much for your solution. Because performance optimization is required for multiple aggregate communications. By extension, as for
`--mca coll_tuned_reduce_scatter_algorithm_tree_fanout and --mca coll_tuned_reduce_scatter_algorithm_chain_fanout` of reduce_scatter.
`--mca coll_tuned_allreduce_algorithm_chain_fanout and --mca coll_tuned_allreduce_algorithm_tree_fanout` of allreduce.
`--mca coll_tuned_reduce_algorithm_tree_fanout and --mca coll_tuned_reduce_algorithm_chain_fanout `of reduce.
`--mca coll_tuned_bcast_algorithm_tree_fanout and --mca coll_tuned_bcast_algorithm_chain_fanout` of bcast
Aren't the parameters actually used? Do you think that's right? Or which of these parameters can be used? If used, what is the value range?

**Comment by gkatev** [comment]
So for other operations yes more parameters are used! Your safest bet is always looking in the implementation. For example for bcast, under ompi/mca/coll/tuned/coll_tuned_bcast_decision.c: https://github.com/open-mpi/ompi/blob/88d8972a4085f6492dc38786c925f9dec446d49e/ompi/mca/coll/tuned/coll_tuned_bcast_decision.c#L134-L170

- Here the range for selection of algorithm is 0 to 9.
- Algorithm 2-9 do not seem to ignore the segsize, so for these you can try various segsizes. It wouldn't make sense to use a segsize that is larger than your message size, so I suggest testing a bunch of sizes from 0 up until a smaller than your message size (if segsize == msgsize, there is no segmenting).
- For Bcast, algorithm 2 (chain) is the only one that makes use of the fan in/out parameter.
  - I'm not entirely sure where `faninout` is initialized and if it gets the value of `_chain_fanout` or `_tree_fanout`, but I expect you will set `--mca coll_tuned_bcast_algorithm_chain_fanout` for the chain algorithm!
- For fanout parameters in general your selection will probably be bounded by the number of ranks. You will probably want to test values between 1-10, or maybe 10-100 but this is just a guess.
  - I'm not sure which algorithm `ompi_coll_base_bcast_intra_chain` implements. Looking at the code is always the safe bet. It might be K-chain tree algorithm, from [here](https://www.sciencedirect.com/science/article/pii/S0743731522000697), fig 3d -- in this example fanout=4, see attached image below. Perhaps @bosilca can chime in with more experience about the specific algorithms.

<details>

![1-s2 0-S0743731522000697-gr004](https://user-images.githubusercontent.com/31203804/217764377-f5a1e3d9-d9ce-4908-9f05-49b12a12ba11.jpg)

</details>

- For bcast also, for algorithm 7 there is the `--mca bcast_algorithm_knomial_radix` parameter. I'd say treat similarly to the fanout one and choose values in the low 1s or 10s. Try to look up "K-nomial broadcast" to see what happens with different K (radix) parameters.

For the rest of the operations the switch logic is in `ompi/mca/coll/tuned/coll_tuned_${OPERATION}_decision.c`, and the algorithms are probably implemented in `ompi/mca/coll/base/coll_base_${OPERATION}.c`.

So to give you a final more straight answer, for the fanout and for the radix params, you can test values of 1, 2, 4, 8, 16, 32, and continue depending on the performance. I don't know how much processes you are testing with, and if they are all on the same node or not. It wouldn't make sense for fanout/radix to be larger than the number of processes you have.

**Comment by gukejun1** [comment]
It's not. every collective (11 types, including alltoall, broadcast, and scatter) The optimal algorithm combination is obtained. I optimized the algorithm for a single collective, a single size, and a single rank number, and traversed the algorithm to find the optimal algorithm combination in each case. So the total time is very long.
